File: src/holo/framing.py
"""Framing protocol for the holo channel.

Frames carry commands (daemon → page) and replies (page → daemon) over any
underlying transport — clipboard paste, document.title, or a same-origin
bridge socket. The wire format is identical regardless of transport, so the
two ends speak the same protocol whether they're co-located or paired
through a registry across the internet.

A frame is JSON with these fields:

    v        protocol version (currently 1)
    session  session id (uuid string), set at calibration
    type     "cmd" | "result" | "ack" | "ping" | "pong" | "bye"
    seq      0-indexed chunk number within this message
    total    total number of chunks in this message (>= 1)
    id       idempotency key (uuid string), shared across chunks of one message
    data     base64-encoded payload bytes (may be empty for ack/ping/bye)
    crc      crc32 of the decoded `data` bytes (8 hex chars)

Corruption is detected by JSON parse failure or CRC mismatch. Re-delivering
the same `id`+`seq` pair is idempotent — the receiver tracks delivered ids
and ignores replays. Page navigation is signalled by a "bye" frame so the
daemon can abort cleanly rather than typing into a dead document.
"""

from __future__ import annotations

import base64
import binascii
import json
import zlib
from dataclasses import dataclass, field
from typing import Literal
from uuid import uuid4

PROTOCOL_VERSION = 1

FrameType = Literal["cmd", "result", "ack", "ping", "pong", "bye"]

_VALID_TYPES: frozenset[str] = frozenset(
    {"cmd", "result", "ack", "ping", "pong", "bye"}
)
_REQUIRED_FIELDS: frozenset[str] = frozenset(
    {"v", "session", "type", "seq", "total", "id", "data", "crc"}
)


class FrameError(Exception):
    """Raised when a frame fails to decode or validate."""


@dataclass(slots=True, frozen=True)
class Frame:
    session: str
    type: FrameType
    data: bytes = b""
    seq: int = 0
    total: int = 1
    id: str = field(default_factory=lambda: str(uuid4()))
    v: int = PROTOCOL_VERSION

    def encode(self) -> str:
        envelope = {
            "v": self.v,
            "session": self.session,
            "type": self.type,
            "seq": self.seq,
            "total": self.total,
            "id": self.id,
            "data": base64.b64encode(self.data).decode("ascii"),
            "crc": format(zlib.crc32(self.data), "08x"),
        }
        return json.dumps(envelope, separators=(",", ":"), sort_keys=True)


def decode(raw: str) -> Frame:
    """Parse and validate a wire-format frame string. Raises FrameError on any defect."""
    try:
        env = json.loads(raw)
    except json.JSONDecodeError as e:
        raise FrameError(f"invalid json: {e}") from e

    if not isinstance(env, dict):
        raise FrameError("frame is not a json object")

    missing = _REQUIRED_FIELDS - env.keys()
    if missing:
        raise FrameError(f"missing fields: {sorted(missing)}")

    if env["v"] != PROTOCOL_VERSION:
        raise FrameError(f"unsupported version: {env['v']}")

    if env["type"] not in _VALID_TYPES:
        raise FrameError(f"unknown frame type: {env['type']}")

    try:
        data = base64.b64decode(env["data"], validate=True)
    except (ValueError, binascii.Error) as e:
        raise FrameError(f"invalid base64 data: {e}") from e

    actual_crc = format(zlib.crc32(data), "08x")
    if actual_crc != env["crc"]:
        raise FrameError(f"crc mismatch: expected {env['crc']}, got {actual_crc}")

    return Frame(
        session=env["session"],
        type=env["type"],
        data=data,
        seq=env["seq"],
        total=env["total"],
        id=env["id"],
        v=env["v"],
    )


def chunk(
    payload: bytes,
    *,
    session: str,
    type: FrameType,
    max_chunk: int = 32 * 1024,
) -> list[Frame]:
    """Split a payload into one or more frames sharing an id, with seq + total set."""
    if max_chunk <= 0:
        raise ValueError("max_chunk must be positive")
    if not payload:
        return [Frame(session=session, type=type, data=b"", total=1)]
    parts = [payload[i : i + max_chunk] for i in range(0, len(payload), max_chunk)]
    msg_id = str(uuid4())
    return [
        Frame(session=session, type=type, data=p, seq=i, total=len(parts), id=msg_id)
        for i, p in enumerate(parts)
    ]


class Reassembler:
    """Collects sequenced frames sharing an id; emits the joined payload when complete.

    Idempotent: re-feeding the same (id, seq) is a no-op once the message is
    complete, and duplicate chunks during reassembly are ignored. Callers
    should track returned non-None payloads as the authoritative delivery
    signal — repeats return None.
    """

    def __init__(self) -> None:
        self._buffers: dict[str, dict[int, bytes]] = {}
        self._totals: dict[str, int] = {}
        self._delivered: set[str] = set()

    def feed(self, frame: Frame) -> bytes | None:
        if frame.id in self._delivered:
            return None
        if not (0 <= frame.seq < frame.total):
            raise FrameError(
                f"seq {frame.seq} out of range for total {frame.total}"
            )
        prior_total = self._totals.get(frame.id)
        if prior_total is not None and prior_total != frame.total:
            raise FrameError(
                f"inconsistent total for id {frame.id}: "
                f"saw {prior_total}, now {frame.total}"
            )
        buf = self._buffers.setdefault(frame.id, {})
        if frame.seq in buf:
            return None
        buf[frame.seq] = frame.data
        self._totals[frame.id] = frame.total
        if len(buf) == frame.total:
            payload = b"".join(buf[i] for i in range(frame.total))
            self._delivered.add(frame.id)
            del self._buffers[frame.id]
            del self._totals[frame.id]
            return payload
        return None

File: bookmarklet/framing.js
// Framing protocol for the holo channel — JS port of holo.framing.
//
// The wire format is identical to the Python implementation: a sorted-key
// JSON envelope with version, session id, type, seq, total, idempotency
// id, base64-encoded data, and a CRC32 of the decoded data bytes.
// Frames produced here decode cleanly on the Python side, and vice versa.
//
// Designed to run in both browsers and Node 20+ (`globalThis.crypto`,
// `atob`, `btoa`, `JSON`, `TextEncoder`). No external dependencies.

export const PROTOCOL_VERSION = 1;

const VALID_TYPES = new Set(["cmd", "result", "ack", "ping", "pong", "bye"]);
const REQUIRED_FIELDS = ["v", "session", "type", "seq", "total", "id", "data", "crc"];

export class FrameError extends Error {
  constructor(message) {
    super(message);
    this.name = "FrameError";
  }
}

// CRC32 (IEEE 802.3, reflected, init/xorout 0xFFFFFFFF). Matches Python's
// zlib.crc32. Polynomial 0xEDB88320 is the reflected form of 0x04C11DB7.
export function crc32(bytes) {
  let crc = 0xffffffff >>> 0;
  for (let i = 0; i < bytes.length; i++) {
    crc = (crc ^ bytes[i]) >>> 0;
    for (let j = 0; j < 8; j++) {
      crc = ((crc >>> 1) ^ (0xedb88320 & -(crc & 1))) >>> 0;
    }
  }
  return ((crc ^ 0xffffffff) >>> 0).toString(16).padStart(8, "0");
}

function bytesToBase64(bytes) {
  // Build a binary string in chunks to avoid the spread-into-fromCharCode
  // limit on very large arrays.
  let s = "";
  const chunk = 0x8000;
  for (let i = 0; i < bytes.length; i += chunk) {
    s += String.fromCharCode.apply(null, bytes.subarray(i, i + chunk));
  }
  return btoa(s);
}

function base64ToBytes(b64) {
  const s = atob(b64);
  const bytes = new Uint8Array(s.length);
  for (let i = 0; i < s.length; i++) bytes[i] = s.charCodeAt(i);
  return bytes;
}

function newId() {
  return globalThis.crypto.randomUUID();
}

function sortedJson(obj) {
  // Sort keys to match Python's json.dumps(..., sort_keys=True). Recursively
  // covers nested objects, although our envelopes are flat.
  if (Array.isArray(obj)) {
    return "[" + obj.map(sortedJson).join(",") + "]";
  }
  if (obj && typeof obj === "object") {
    const keys = Object.keys(obj).sort();
    const parts = keys.map((k) => JSON.stringify(k) + ":" + sortedJson(obj[k]));
    return "{" + parts.join(",") + "}";
  }
  return JSON.stringify(obj);
}

export function encodeFrame({
  session,
  type,
  data = new Uint8Array(),
  seq = 0,
  total = 1,
  id = newId(),
  v = PROTOCOL_VERSION,
}) {
  const envelope = {
    v,
    session,
    type,
    seq,
    total,
    id,
    data: bytesToBase64(data),
    crc: crc32(data),
  };
  return sortedJson(envelope);
}

export function decodeFrame(raw) {
  let env;
  try {
    env = JSON.parse(raw);
  } catch (e) {
    throw new FrameError(`invalid json: ${e.message}`);
  }
  if (env === null || typeof env !== "object" || Array.isArray(env)) {
    throw new FrameError("frame is not a json object");
  }
  const missing = REQUIRED_FIELDS.filter((k) => !(k in env));
  if (missing.length > 0) {
    throw new FrameError(`missing fields: ${missing.sort().join(",")}`);
  }
  if (env.v !== PROTOCOL_VERSION) {
    throw new FrameError(`unsupported version: ${env.v}`);
  }
  if (!VALID_TYPES.has(env.type)) {
    throw new FrameError(`unknown frame type: ${env.type}`);
  }
  let data;
  try {
    data = base64ToBytes(env.data);
  } catch (e) {
    throw new FrameError(`invalid base64 data: ${e.message}`);
  }
  const actualCrc = crc32(data);
  if (actualCrc !== env.crc) {
    throw new FrameError(`crc mismatch: expected ${env.crc}, got ${actualCrc}`);
  }
  return {
    v: env.v,
    session: env.session,
    type: env.type,
    seq: env.seq,
    total: env.total,
    id: env.id,
    data,
  };
}

export function chunkPayload(payload, { session, type, maxChunk = 32 * 1024 }) {
  if (!(payload instanceof Uint8Array)) {
    throw new TypeError("payload must be a Uint8Array");
  }
  if (maxChunk <= 0) {
    throw new RangeError("maxChunk must be positive");
  }
  if (payload.length === 0) {
    return [{ session, type, data: new Uint8Array(), seq: 0, total: 1, id: newId() }];
  }
  const id = newId();
  const total = Math.ceil(payload.length / maxChunk);
  const frames = [];
  for (let i = 0, seq = 0; i < payload.length; i += maxChunk, seq++) {
    frames.push({
      session,
      type,
      data: payload.subarray(i, Math.min(i + maxChunk, payload.length)),
      seq,
      total,
      id,
    });
  }
  return frames;
}

export class Reassembler {
  constructor() {
    this._buffers = new Map();
    this._totals = new Map();
    this._delivered = new Set();
  }

  feed(frame) {
    if (this._delivered.has(frame.id)) return null;
    if (!(frame.seq >= 0 && frame.seq < frame.total)) {
      throw new FrameError(`seq ${frame.seq} out of range for total ${frame.total}`);
    }
    const priorTotal = this._totals.get(frame.id);
    if (priorTotal !== undefined && priorTotal !== frame.total) {
      throw new FrameError(
        `inconsistent total for id ${frame.id}: saw ${priorTotal}, now ${frame.total}`
      );
    }
    let buf = this._buffers.get(frame.id);
    if (!buf) {
      buf = new Map();
      this._buffers.set(frame.id, buf);
    }
    if (buf.has(frame.seq)) return null;
    buf.set(frame.seq, frame.data);
    this._totals.set(frame.id, frame.total);
    if (buf.size === frame.total) {
      let totalLen = 0;
      for (let i = 0; i < frame.total; i++) totalLen += buf.get(i).length;
      const payload = new Uint8Array(totalLen);
      let offset = 0;
      for (let i = 0; i < frame.total; i++) {
        const part = buf.get(i);
        payload.set(part, offset);
        offset += part.length;
      }
      this._delivered.add(frame.id);
      this._buffers.delete(frame.id);
      this._totals.delete(frame.id);
      return payload;
    }
    return null;
  }
}
